src/UI/Layout.hs

```diff
@@ -76,36 +103,29 @@ layoutRectanglesAlgebra :: Real a => Algebra (Fitting (LayoutF a) a) [Rect a]
 layoutRectanglesAlgebra = wrapAlgebra catMaybes (fmap Just) (collect layoutAlgebra)
 
 
-type Fitting f a = Bidi f (Size a) (Point a, Size (Maybe a))
+type Fitting f a = Bidi f (Size a) (Alignment, Point a, Size (Maybe a))
 
 fitLayoutWith :: Real a => Algebra (Fitting (LayoutF a) a) b -> Size (Maybe a) -> Layout a (Size a) -> b
-fitLayoutWith algebra maxSize layout = hylo algebra fittingCoalgebra (Point 0 0, maxSize, layout)
+fitLayoutWith algebra maxSize layout = hylo algebra fittingCoalgebra (Full, Point 0 0, maxSize, layout)
 
-fittingCoalgebra :: Real a => Coalgebra (Fitting (LayoutF a) a) (Point a, Size (Maybe a), Layout a (Size a))
-fittingCoalgebra (offset, maxSize, layout) = Cofree (offset, maxSize) id $ case runFreer layout of
+fittingCoalgebra :: Real a => Coalgebra (Fitting (LayoutF a) a) (Alignment, Point a, Size (Maybe a), Layout a (Size a))
+fittingCoalgebra (alignment, offset, maxSize, layout) = Cofree (alignment, offset, maxSize) id $ case runFreer layout of
   Pure size -> Pure size
   Free run layout -> case layout of
-    Inset by child -> Free id $ Inset by (addSizeToPoint offset by, subtractSize maxSize (2 * by), run child)
-    Offset by child -> Free id $ Offset by (liftA2 (+) offset by, subtractSize maxSize (pointSize by), run child)
-    GetMaxSize -> Free ((,,) offset maxSize . run) GetMaxSize
+    Inset by child -> Free id $ Inset by (alignment, addSizeToPoint offset by, subtractSize maxSize (2 * by), run child)
+    Offset by child -> Free id $ Offset by (alignment, liftA2 (+) offset by, subtractSize maxSize (pointSize by), run child)
+    GetMaxSize -> Free ((,,,) alignment offset maxSize . run) GetMaxSize
+    Align alignment child -> Free id $ Align alignment (alignment, offset, maxSize, run child)
   where subtractSize maxSize size = liftA2 (-) <$> maxSize <*> (Just <$> size)
-        addSizeToPoint point (Size w h) = liftA2 (+) point (Point w h)
+        addSizeToPoint point = liftA2 (+) point . sizeExtent
 
 
 -- Instances
 
-instance Foldable (LayoutF a) where
-  foldMap f layout = case layout of
-    Inset _ child -> f child
-    Offset _ child -> f child
-    GetMaxSize -> f (pure Nothing)
+instance Real a => Semigroup (Layout a (Size a)) where
+  (<>) = stack
```

Comment:
It also turns out that stacking is a great `Semigroup` instance. The intention is to make adjacency the `Alternative` instance.